frontend/packages/console-dynamic-plugin-sdk/src/utils/k8s/k8s-ref.ts

```diff
@@ -0,0 +1,77 @@
+import { K8sModel } from '../../api/common-types';
+import {
+  K8sResourceKindReference,
+  K8sGroupVersionKind,
+  K8sResourceCommon,
+} from '../../extensions/console-types';
+
+/**
+ * Provides a reference string that uniquely identifies the group, version, and kind of a k8s resource.
+ * @param group Pass group of k8s resource or model.
+ * @param version Pass version of k8s resource or model.
+ * @param kind Pass kind of k8s resource or model.
+ * @return The reference for any k8s resource i.e `group~version~kind`.
+ * * */
+export const getReference = (
+  group: string,
+  version: string,
+  kind: string,
+): K8sResourceKindReference => [group, version, kind].join('~');
+
+/**
+ * Provides a reference string that uniquely identifies the group, version, and kind of a k8s model.
+ * @param model k8s model
+ * @return The reference for model i.e `group~version~kind`.
+ * * */
+export const getReferenceForModel = (model: K8sModel): K8sResourceKindReference =>
+  getReference(model.apiGroup || 'core', model.apiVersion, model.kind);
+
+/**
+ * Provides apiVersion for a k8s model.
+ * @param model k8s model
+ * @return The apiVersion for the model i.e `group/version`.
+ * * */
+export const getAPIVersionForModel = (model: K8sModel): string =>
+  !model?.apiGroup ? model.apiVersion : `${model.apiGroup}/${model.apiVersion}`;
+
+/**
+ * Provides a group, version, and kind for a resource.
+ * @param resource k8s resource
+ * @return The group, version, kind for the provided resource.
+ * If the resource does not have an API group, group "core" will be returned.
+ * If the resource has an invalid apiVersion then null will be returned.
+ * * */
+export const getGroupVersionKindForResource = (
+  resource: K8sResourceCommon,
+): K8sGroupVersionKind => {
+  const { apiVersion, kind } = resource;
+  const apiVersionSplit = apiVersion.split('/');
+  const apiVersionSplitLen = apiVersionSplit.length;
+  if (apiVersionSplitLen > 2) {
+    // eslint-disable-next-line no-console
+    console.log('Provided resource has invalid apiVersion.');
+    return null;
+  }
+  return {
+    group: apiVersionSplitLen === 2 ? apiVersionSplit[0] : 'core',
+    version: apiVersionSplitLen === 2 ? apiVersionSplit[1] : apiVersion,
+    kind,
+  };
+};
+
+/**
+ * Provides a group, version, and kind for a reference.
+ * @param reference reference for group, version, kind i.e `group~version~kind`.
+ * @return The group, version, kind for the provided reference.
+ * If the group's value is "core" it denotes resource does not have an API group.
+ * * */
+export const getGroupVersionKindForReference = (
+  reference: K8sResourceKindReference,
+): K8sGroupVersionKind => {
+  const [group, version, kind] = reference.split('~');
```

Comment:
We should add error handling as well here in case this isn't a valid reference string.